Sources/CountedSet.swift

```diff
@@ -245,3 +259,50 @@ extension CountedSet: Sequence {
 		return backingDictionary.makeIterator()
 	}
 }
+
+extension CountedSet {
+    public func reduce<Result>(_ initialResult: Result, _ nextPartialResult: (Result, (element: Element, count:Int)) throws -> Result) rethrows -> Result {
+        return try backingDictionary.reduce(initialResult) { (result, info) in
+            try nextPartialResult(result, (element: info.key, count: info.value))
+        }
+    }
+    
+    public func map<U>(_ transform: (Element, Int) throws -> (element:U, count:Int)) rethrows -> CountedSet<U> {
+        var result = CountedSet<U>()
+        
+        for (element, count) in backingDictionary {
+            let (newElement, newCount) = try transform(element, count)
+            result.update(with: newElement, count: newCount)
+        }
+    
+        return result
+    }
+
+    public func flatMap<U>(_ transform: (Element, Int) throws -> (element:U, count:Int)?) rethrows -> CountedSet<U> {
+        var result = CountedSet<U>()
+        
+        for (element, count) in backingDictionary {
+            if let (newElement, newCount) = try transform(element, count) {
+                result.update(with: newElement, count: newCount)
+            }
+        }
+        
+        return result
+    }
+
+    public func filter(_ inclusionCount: (Element, Int) throws -> Int) rethrows -> CountedSet<Element> {
```

Comment:
The original `filter` function returns a boolean to include or not the element in the resulting collection.  As the `CountedSet` contains the number of element, the idea of the filter is to return the new count of the element, not only its presence. 

But as I commented, maybe we could just rename if `countedSetFilter`?